### PR TITLE
Add custom grains containing the redis elasticsearch endpoints

### PR DESCRIPTION
**Pull request by sevenmachines**
Sets grains for elasticache endpoints for a redis type. Consisting of
a primary (read-write) endpoint and a list of read-only endpoints.
These endpoints will only be collected if there is already a grain
ElasticacheReplicationGroupName containing the replication group
name.

The grain set has the form,

```
  'elasticache':  {
    'primary_endpoint': {'address': <address>, 'port': <port>},
    'default_endpoint': {'address': <address>, 'port': <port>},
    'read_endpoints': [
      {'address': <address>, 'port': <port>},
      {'address': <address>, 'port': <port>},
      {'address': <address>, 'port': <port>}
    ]
  }
```

Requires PR https://github.com/ministryofjustice/bootstrap-cfn/pull/152
And note issue: https://github.com/ministryofjustice/template-deploy/issues/171

### CODE REVIEW

**Comment by sevenmachines** [comment]
Looks good

```
# salt-call grains.get elasticache

local:
    ----------
    default_endpoint:
        ----------
        Address:
            teodg9rfj1g1.eq1onc.ng.0001.euw1.cache.amazonaws.com
        Port:
            10024
    primary_endpoint:
        ----------
        Address:
            teodg9rfj1g1.eq1onc.ng.0001.euw1.cache.amazonaws.com
        Port:
            10024
    read_endpoints:
        |_
          ----------
          Address:
              teodg9rfj1g1-001.eq1onc.0001.euw1.cache.amazonaws.com
          Port:
              10024
        |_
          ----------
          Address:
              teodg9rfj1g1-002.eq1onc.0001.euw1.cache.amazonaws.com
          Port:
              10024
        |_
          ----------
          Address:
              teodg9rfj1g1-003.eq1onc.0001.euw1.cache.amazonaws.com
          Port:
              10024
```

**Comment by sevenmachines** [comment]
Theres a duplication in code from tags_to_grains to get the ElasticacheReplicationGroupName output from AWS. Is there a better way to do this? (partially rhetorical :) )

**Comment by sevenmachines** [comment]
Security groups seem ok, from ec2 instance

```
root@i-e0075241:~# telnet teeodg9rfoj1gp1.eq1onc.ng.0001.euw1.cache.amazonaws.com 10024
Trying 10.0.27.124...
Connected to teeodg9rfoj1gp1-001.eq1onc.0001.euw1.cache.amazonaws.com.
```

**Comment by sevenmachines** [comment]
Might want to add a check for status:creating and warn(?), we arent going to get endpoints until status:available

**Comment by sevenmachines** [comment]
Other points worth thinking about...
- defaults for replication groups
- what ENVs to expose? have REDIS_HOST of primary endpoint, maybe want RO endpoints as well?

**Comment by andyhd** [comment]
Not sure I can comment on the functionality, but I have a lot to say about the Python code. Do you want my comments?

**Comment by ashb** [comment]
Can you also set REDIS_URL as a single env var? A number of (ruby) libs will look for this and use it by default where they won't the HOST and PORT version.

File: _grains/elasticache.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    metadata_ip = '169.254.169.254'
+    try:
+        instance_id = urllib2.urlopen(
+            'http://{0}/latest/meta-data/instance-id'.format(metadata_ip)).read().strip()
+        region = urllib2.urlopen(
+            'http://{0}/latest/meta-data/placement/availability-zone'.format(metadata_ip)).read().strip()[:-1]
+        return {'aws_instance_id': instance_id, 'aws_region': region}
+    except Exception as err:
+        sys.stderr.write('tags_to_grains ERROR: %s\n' % str(err))
+        return {'custom_grain_error': True}
```
Comment (ashb) [inline _grains/elasticache.py]: Boto has native support for querying the metadata API - we hould use that rather than making the requests ourselves.

Comment (ashb) [inline _grains/elasticache.py]: http://boto.readthedocs.org/en/latest/ref/boto.html?highlight=metadata#boto.utils.get_instance_metadata

http://boto.readthedocs.org/en/latest/ref/boto.html?highlight=metadata#boto.utils.get_aws_metadata

**Comment by ashb** [comment]
So also: the custom grain is in the wrong repo - this shoud live in aws-formula or alongside https://github.com/ministryofjustice/bootstrap-salt/blob/master/bootstrap_salt/contrib/srv/salt/_grains/tags_to_grains.py in that file.

File: _grains/elasticache.py (hunk 1 above)
Comment (sevenmachines) [inline _grains/elasticache.py]: The dangers of copy/paste, I've also left 'courtfinder' in a function comment :)

**Comment by sevenmachines** [comment]
@andyhd all comments welcome!

**Comment by sevenmachines** [comment]
@ashb I'll add REDIS_URL and boto-ise as mentioned. I'll move the grain to aws-formula I think

**Comment by sevenmachines** [comment]
Waiting on aws-formula merge release of https://github.com/ministryofjustice/aws-formula/pull/3

**Comment by sevenmachines** [comment]
aws-formula 0.5.0 released so this should be good to go

**Comment by pwyborn** [comment]
@niallcreech - I don't think that I have "write access" to this repository (not giving merge option). This looks good to go though (works for me in my app)

**Comment by sevenmachines** [comment]
@bennythejudge Should be good now

**Comment by bennythejudge** [comment]
I am reviewing this

**Pull request merged**